src/symbolTable.ts

```diff
@@ -19,59 +19,181 @@ export default class SymbolTable {
+    throw new Error("Void isn't iterable");
+  }
+}
+
+export class UndefinedType extends Type {
+  public toString = (): string => { return "undefined"};
 
-export class StringType extends Type {}
+  public isIterable(): boolean {
+    return false;
+  }
 
-export class BooleanType extends Type {}
+  public getSpreadTypes(): Type[] {
+    throw new Error("Undefined isn't iterable");
+  }
+}
 
-export class VoidType extends Type {}
+export class NullType extends Type {
+  public toString = (): string => { return "null"};
 
-export class UndefinedType extends Type {}
+  public isIterable(): boolean {
+    return false;
+  }
 
-export class NullType extends Type {}
+  public getSpreadTypes(): Type[] {
+    throw new Error("Null isn't iterable");
+  }
+}
 
 // compound types
 export class ObjectType extends Type {
   public fields: [string | number, Type][] = [];
+  public toString = (): string => {
+    return `object with fields: ${this.fields}`;
+  };
 
   constructor(fields: [string | number, Type][]) {
     super();
     this.fields = fields;
   }
+
+  public isIterable(): boolean {
+    return true;
+  }
+
+  public getSpreadTypes(): Type[] {
+    return this.fields.map(value => {return value[1]});
+  }
 }
 
 export class ArrayType extends Type {
   public elementTypes: Type[];
+  public toString = (): string => {
+    return `array of ${this.elementTypes}`;
+  };
 
   constructor(elementTypes: Type[]) {
     super();
     this.elementTypes = elementTypes;
   }
+
+  public isIterable(): boolean {
+    return true;
+  }
+
+  public getSpreadTypes(): Type[] {
+    return this.elementTypes;
+  }
 }
 
 export class FunctionType extends Type {
   public params: Type[];
   public returnType: Type;
+  public toString = (): string => {
+    return `function with parameter types ${this.params} and return type ${this.returnType}`;
+  };
 
   constructor(params: Type[], returnType: Type) {
     super();
     this.params = params;
     this.returnType = returnType;
   }
+
+  public isIterable(): boolean {
+    return false;
+  }
+
+  public getSpreadTypes(): Type[] {
+    throw new Error("Function isn't iterable");
+  }
 }
 
 // computed types
 export class UnionType extends Type {
   public types: Type[];
+  public toString = (): string => {
+    return `one of the following types ${this.types}`;
+  };
 
   constructor(types: Type[]) {
     super();
     this.types = types;
   }
+
+  public isIterable(): boolean {
+    return this.types.filter(type => type.isIterable()).length != 0;
+  }
+
+  public getSpreadTypes(): Type[] {
+    this.types = this.types.filter(type => type.isIterable());
+    return [this];
+  }
 }
 
-export class AnyType extends Type {}
+export class AnyType extends Type {
+  public toString = (): string => {
+    return "any type";
+  };
+
+  public isIterable(): boolean {
+    return true;
+  }
+
+  public getSpreadTypes(): Type[] {
+    return [new UnionType([new StringType(), new ArrayType([new AnyType()]), new ObjectType([])])];
```

Comment:
Is there any case where this will return two or more Type objects that can't be encapsulated together with a single UnionType? It might also be useful to add a default implementation for this function that raises an error by default.